node.c

```diff
@@ -483,7 +523,32 @@ rb_ast_delete_mark_object(rb_ast_t *ast, VALUE obj)
 VALUE
 rb_ast_tokens(rb_ast_t *ast)
 {
-    return ast->node_buffer->tokens;
+    long i;
```

Comment:
How about to move this ruby object construction logic to `ast.c`? By the change, node.c no need to handle VALUE.

Reply:
Yeah, I thought that too but,
If we move the logic from node.c to ast.c, do you also think we should redesign `ast->node_buffer->tokens` into `ast->tokens` because we can say the `tokens` member no longer belongs to a Node?
(Of course, we cannot change the Ruby level design: `AbstractSyntaxTree::Node` though)

Reply:
In a long term, it's correct however It's impossible right now. `rb_ast_t` is imemo (internal ruby object) now therefore it's size should be 5 words. It's already uses 5 words like below. We can refactor it once `rb_ast_t` is not managed as `imemo`.

```c
typedef struct rb_ast_body_struct {
    const NODE *root;
    VALUE script_lines;
    // script_lines is either:
    // - a Fixnum that represents the line count of the original source, or
    // - an Array that contains the lines of the original source
    signed int frozen_string_literal:2; /* -1: not specified, 0: false, 1: true */
    signed int coverage_enabled:2; /* -1: not specified, 0: false, 1: true */
} rb_ast_body_t;
typedef struct rb_ast_struct {
    VALUE flags;
    node_buffer_t *node_buffer;
    rb_ast_body_t body;
} rb_ast_t;
```

Reply:
Ah, I see

Reply:
I did
https://github.com/hasumikin/ruby/commit/5ef62bbdb1c1a7dd01e5730246b4f2f2ff8cad94